numpy/core/src/multiarray/convert_datatype.c

```diff
@@ -262,9 +259,16 @@ PyArray_AdaptFlexibleDType(PyObject *data_obj, PyArray_Descr *data_dtype,
                             list = PyArray_ToList((PyArrayObject *)data_obj);
                             result = PyArray_GetArrayParamsFromObject(
                                     list,
-                                    *flex_dtype,
+                                    retval,
                                     0, &dtype,
                                     &ndim, dims, &arr, NULL);
+                            if (result < 0) {
+                                Py_XDECREF(dtype);
+                                Py_XDECREF(arr);
+                                Py_DECREF(list);
```

Comment:
Might it be nicer to DECREF arr and list uncoditionally before and remove it from the end?

Reply:
adopted